File: Assignment-4/python/schedule_manager.py
#!/usr/bin/env python3
"""
Employee Schedule Manager - Python Implementation
Manages employee schedules with morning, afternoon, and evening shifts across 7 days.
Uses conditionals, loops, and branching to implement scheduling logic.
"""

import random
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass, field

# Constants
DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["morning", "afternoon", "evening"]
MIN_EMPLOYEES_PER_SHIFT = 2
MAX_DAYS_PER_EMPLOYEE = 5
FIRST_PASS_MAX_DAYS = 4  # Leave capacity for backfill to meet min 2 per shift
MAX_PER_SHIFT = 4  # Cap per shift to allow conflict detection ("full")


@dataclass
class Employee:
    """Stores employee name and shift preferences per day."""
    name: str
    # preferences[day_index] = preferred shift name, or list for priority (bonus)
    preferences: Dict[int, str] = field(default_factory=dict)
    # Optional: priority_ranking[day_index] = [shift1, shift2, shift3] (1st, 2nd, 3rd choice)
    priority_ranking: Optional[Dict[int, List[str]]] = None

    def get_preferred_shift(self, day: int) -> str:
        """Return preferred shift for day (single preference)."""
        return self.preferences.get(day, random.choice(SHIFTS))

    def get_priority_list(self, day: int) -> List[str]:
        """Return ordered list of shift preferences for day (bonus)."""
        if self.priority_ranking and day in self.priority_ranking:
            return self.priority_ranking[day]
        default = [self.get_preferred_shift(day)]
        for s in SHIFTS:
            if s not in default:
                default.append(s)
        return default


def build_schedule_structure() -> Dict[int, Dict[str, List[str]]]:
    """Create empty schedule: schedule[day_index][shift] = list of employee names."""
    schedule = {}
    for day in range(7):
        schedule[day] = {shift: [] for shift in SHIFTS}
    return schedule


def count_employee_days(assignments: List[Tuple[int, str]], employee: str) -> int:
    """Count how many days an employee is assigned."""
    return sum(1 for d, _ in assignments if d is not None)


def get_employee_assignments(schedule: Dict[int, Dict[str, List[str]]]) -> Dict[str, List[Tuple[int, str]]]:
    """From schedule, build map: employee_name -> [(day_index, shift), ...]."""
    result = {}
    for day in range(7):
        for shift in SHIFTS:
            for name in schedule[day][shift]:
                if name not in result:
                    result[name] = []
                result[name].append((day, shift))
    return result


def is_shift_full(schedule: Dict[int, Dict[str, List[str]]], day: int, shift: str) -> bool:
    """Return True if this shift on this day has reached capacity (for conflict)."""
    return len(schedule[day][shift]) >= MAX_PER_SHIFT


def days_worked(employee_assignments: Dict[str, List[Tuple[int, str]]], name: str) -> int:
    """Return number of days the employee is assigned."""
    return len(employee_assignments.get(name, []))


def first_pass_assign(
    employees: List[Employee],
    schedule: Dict[int, Dict[str, List[str]]],
    use_priority: bool,
) -> None:
    """
    Assign employees to shifts by preference (or priority order).
    No more than one shift per day, max 5 days per employee.
    Resolve conflicts: if preferred shift full, assign to another shift same day or next day.
    """
    employee_assignments: Dict[str, List[Tuple[int, str]]] = {}

    day_order = list(range(7))
    for emp in employees:
        days_assigned = 0
        random.shuffle(day_order)  # Spread assignments across week (including weekend)
        for day in day_order:
            if days_assigned >= FIRST_PASS_MAX_DAYS:
                break
            if day in [a[0] for a in employee_assignments.get(emp.name, [])]:
                continue  # Already assigned this day

            if use_priority and emp.priority_ranking and day in emp.priority_ranking:
                shift_order = emp.priority_ranking[day]
            else:
                preferred = emp.get_preferred_shift(day)
                shift_order = [preferred] + [s for s in SHIFTS if s != preferred]

            assigned = False
            for shift in shift_order:
                if is_shift_full(schedule, day, shift):
                    continue  # Conflict: shift full, try next preference
                schedule[day][shift].append(emp.name)
                employee_assignments.setdefault(emp.name, []).append((day, shift))
                days_assigned += 1
                assigned = True
                break

            if not assigned:
                # Try next day for any available shift (conflict resolution)
                for next_day in range(day + 1, min(day + 2, 7)):
                    if days_assigned >= FIRST_PASS_MAX_DAYS:
                        break
                    if next_day in [a[0] for a in employee_assignments.get(emp.name, [])]:
                        continue
                    for shift in SHIFTS:
                        if not is_shift_full(schedule, next_day, shift):
                            schedule[next_day][shift].append(emp.name)
                            employee_assignments.setdefault(emp.name, []).append((next_day, shift))
                            days_assigned += 1
                            assigned = True
                            break
                    if assigned:
                        break


def backfill_shifts(
    employees: List[Employee],
    schedule: Dict[int, Dict[str, List[str]]],
) -> None:
    """
    Ensure at least MIN_EMPLOYEES_PER_SHIFT per shift per day.
    Randomly assign additional employees who have not worked 5 days yet.
    """
    names = [e.name for e in employees]
    added = True
    while added:
        added = False
        employee_assignments = get_employee_assignments(schedule)
        for day in range(7):
            for shift in SHIFTS:
                current = len(schedule[day][shift])
                needed = max(0, MIN_EMPLOYEES_PER_SHIFT - current)
                if needed == 0:
                    continue

                # Candidates: not on this shift this day, and have worked < 5 days
                candidates = [
                    n for n in names
                    if n not in schedule[day][shift]
                    and days_worked(employee_assignments, n) < MAX_DAYS_PER_EMPLOYEE
                    and not any(a[0] == day for a in employee_assignments.get(n, []))
                ]
                random.shuffle(candidates)
                for i in range(min(needed, len(candidates))):
                    name = candidates[i]
                    schedule[day][shift].append(name)
                    employee_assignments.setdefault(name, []).append((day, shift))
                    added = True
                    break


def collect_input() -> List[Employee]:
    """
    Collect employee names and shift preferences (and optional priority ranking).
    Uses sample data for demo; can be replaced with interactive input.
    """
    employees = []

    # Sample data with preferences per day and optional priority (bonus)
    sample = [
        {
            "name": "Alice",
            "preferences": {0: "morning", 1: "morning", 2: "afternoon", 3: "evening", 4: "morning", 5: "afternoon", 6: "evening"},
            "priority": {0: ["morning", "afternoon", "evening"], 1: ["morning", "evening", "afternoon"], 2: ["afternoon", "morning", "evening"], 3: ["evening", "afternoon", "morning"], 4: ["morning", "afternoon", "evening"], 5: ["afternoon", "morning", "evening"], 6: ["evening", "morning", "afternoon"]},
        },
        {
            "name": "Bob",
            "preferences": {0: "afternoon", 1: "evening", 2: "morning", 3: "morning", 4: "afternoon", 5: "evening", 6: "morning"},
            "priority": {d: ["afternoon", "evening", "morning"] if d % 3 == 0 else ["evening", "morning", "afternoon"] if d % 3 == 1 else ["morning", "afternoon", "evening"] for d in range(7)},
        },
        {
            "name": "Carol",
            "preferences": {0: "evening", 1: "afternoon", 2: "evening", 3: "afternoon", 4: "evening", 5: "morning", 6: "afternoon"},
            "priority": {d: ["evening", "afternoon", "morning"] for d in range(7)},
        },
        {
            "name": "David",
            "preferences": {0: "morning", 1: "afternoon", 2: "morning", 3: "evening", 4: "afternoon", 5: "morning", 6: "evening"},
            "priority": None,
        },
        {
            "name": "Eve",
            "preferences": {0: "afternoon", 1: "morning", 2: "evening", 3: "morning", 4: "evening", 5: "afternoon", 6: "morning"},
            "priority": {d: ["afternoon", "morning", "evening"] for d in range(7)},
        },
        {
            "name": "Frank",
            "preferences": {0: "morning", 1: "evening", 2: "afternoon", 3: "morning", 4: "morning", 5: "evening", 6: "afternoon"},
            "priority": None,
        },
        {
            "name": "Grace",
            "preferences": {0: "evening", 1: "afternoon", 2: "morning", 3: "evening", 4: "afternoon", 5: "morning", 6: "evening"},
            "priority": {d: ["evening", "afternoon", "morning"] for d in range(7)},
        },
        {
            "name": "Henry",
            "preferences": {0: "afternoon", 1: "morning", 2: "evening", 3: "afternoon", 4: "evening", 5: "afternoon", 6: "morning"},
            "priority": None,
        },
        {
            "name": "Ivy",
            "preferences": {0: "morning", 1: "evening", 2: "morning", 3: "afternoon", 4: "afternoon", 5: "morning", 6: "evening"},
            "priority": {d: ["morning", "evening", "afternoon"] for d in range(7)},
        },
        {
            "name": "Jack",
            "preferences": {0: "evening", 1: "morning", 2: "afternoon", 3: "evening", 4: "morning", 5: "afternoon", 6: "morning"},
            "priority": None,
        },
        {
            "name": "Kate",
            "preferences": {0: "afternoon", 1: "evening", 2: "morning", 3: "morning", 4: "evening", 5: "evening", 6: "afternoon"},
            "priority": {d: ["afternoon", "evening", "morning"] for d in range(7)},
        },
    ]

    for s in sample:
        emp = Employee(
            name=s["name"],
            preferences=s["preferences"],
            priority_ranking=s.get("priority"),
        )
        employees.append(emp)

    return employees


def print_schedule(schedule: Dict[int, Dict[str, List[str]]]) -> None:
    """Output the final schedule in a readable format."""
    print("\n" + "=" * 70)
    print("                    FINAL EMPLOYEE SCHEDULE - WEEK")
    print("=" * 70)
    for day in range(7):
        print(f"\n  {DAYS[day].upper()}")
        print("  " + "-" * 60)
        for shift in SHIFTS:
            names = schedule[day][shift]
            names_str = ", ".join(names) if names else "(none)"
            print(f"    {shift.capitalize():10} : {names_str}")
    print("\n" + "=" * 70)


def main() -> None:
    """Run the schedule manager: input, schedule, resolve conflicts, backfill, output."""
    random.seed(42)
    employees = collect_input()
    schedule = build_schedule_structure()

    # Scheduling: first pass with priority preferences (bonus)
    use_priority = any(e.priority_ranking for e in employees)
    first_pass_assign(employees, schedule, use_priority=use_priority)

    # Ensure minimum 2 per shift
    backfill_shifts(employees, schedule)

    print_schedule(schedule)


if __name__ == "__main__":
    main()

File: Assignment-4/java/ScheduleManager.java
import java.util.*;

/**
 * Employee Schedule Manager - Java Implementation
 * Manages employee schedules with morning, afternoon, and evening shifts across 7 days.
 * Uses conditionals, loops, and branching to implement scheduling logic.
 */
public class ScheduleManager {

    private static final String[] DAYS = {
        "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
    };
    private static final String[] SHIFTS = { "morning", "afternoon", "evening" };
    private static final int MIN_EMPLOYEES_PER_SHIFT = 2;
    private static final int MAX_DAYS_PER_EMPLOYEE = 5;
    private static final int FIRST_PASS_MAX_DAYS = 4;
    private static final int MAX_PER_SHIFT = 4;

    private static final Random RANDOM = new Random(42);

    /** schedule[day][shift] = list of employee names */
    private List<List<List<String>>> schedule;
    /** employee name -> list of (day, shift) assignments */
    private Map<String, List<int[]>> employeeAssignments;

    static class Employee {
        String name;
        Map<Integer, String> preferences;
        Map<Integer, List<String>> priorityRanking;

        Employee(String name, Map<Integer, String> preferences, Map<Integer, List<String>> priorityRanking) {
            this.name = name;
            this.preferences = preferences;
            this.priorityRanking = priorityRanking;
        }

        String getPreferredShift(int day) {
            if (preferences != null && preferences.containsKey(day)) {
                return preferences.get(day);
            }
            return SHIFTS[RANDOM.nextInt(SHIFTS.length)];
        }

        List<String> getPriorityList(int day) {
            if (priorityRanking != null && priorityRanking.containsKey(day)) {
                return priorityRanking.get(day);
            }
            List<String> list = new ArrayList<>();
            list.add(getPreferredShift(day));
            for (String s : SHIFTS) {
                if (!list.contains(s)) list.add(s);
            }
            return list;
        }
    }

    public ScheduleManager() {
        schedule = new ArrayList<>(7);
        for (int day = 0; day < 7; day++) {
            List<List<String>> dayShifts = new ArrayList<>(3);
            for (int s = 0; s < 3; s++) {
                dayShifts.add(new ArrayList<>());
            }
            schedule.add(dayShifts);
        }
        employeeAssignments = new HashMap<>();
    }

    private int shiftIndex(String shift) {
        for (int i = 0; i < SHIFTS.length; i++) {
            if (SHIFTS[i].equals(shift)) return i;
        }
        return 0;
    }

    private void buildEmployeeAssignmentsFromSchedule() {
        employeeAssignments.clear();
        for (int day = 0; day < 7; day++) {
            for (int si = 0; si < SHIFTS.length; si++) {
                for (String name : schedule.get(day).get(si)) {
                    employeeAssignments.computeIfAbsent(name, k -> new ArrayList<>())
                        .add(new int[] { day, si });
                }
            }
        }
    }

    private boolean isShiftFull(int day, String shift) {
        int si = shiftIndex(shift);
        return schedule.get(day).get(si).size() >= MAX_PER_SHIFT;
    }

    private int daysWorked(String name) {
        return employeeAssignments.getOrDefault(name, Collections.emptyList()).size();
    }

    private boolean isAssignedOnDay(String name, int day) {
        for (int[] a : employeeAssignments.getOrDefault(name, Collections.emptyList())) {
            if (a[0] == day) return true;
        }
        return false;
    }

    private void firstPassAssign(List<Employee> employees, boolean usePriority) {
        employeeAssignments.clear();

        List<Integer> dayOrder = new ArrayList<>();
        for (int d = 0; d < 7; d++) dayOrder.add(d);

        for (Employee emp : employees) {
            int daysAssigned = 0;
            Collections.shuffle(dayOrder, RANDOM);

            for (int day : dayOrder) {
                if (daysAssigned >= FIRST_PASS_MAX_DAYS) break;
                if (isAssignedOnDay(emp.name, day)) continue;

                List<String> shiftOrder;
                if (usePriority && emp.priorityRanking != null && emp.priorityRanking.containsKey(day)) {
                    shiftOrder = emp.getPriorityList(day);
                } else {
                    String preferred = emp.getPreferredShift(day);
                    shiftOrder = new ArrayList<>();
                    shiftOrder.add(preferred);
                    for (String s : SHIFTS) {
                        if (!s.equals(preferred)) shiftOrder.add(s);
                    }
                }

                boolean assigned = false;
                for (String shift : shiftOrder) {
                    if (isShiftFull(day, shift)) continue;
                    int si = shiftIndex(shift);
                    schedule.get(day).get(si).add(emp.name);
                    employeeAssignments.computeIfAbsent(emp.name, k -> new ArrayList<>())
                        .add(new int[] { day, si });
                    daysAssigned++;
                    assigned = true;
                    break;
                }

                if (!assigned) {
                    int nextDay = day + 1;
                    if (nextDay < 7 && daysAssigned < FIRST_PASS_MAX_DAYS && !isAssignedOnDay(emp.name, nextDay)) {
                        for (String shift : SHIFTS) {
                            if (!isShiftFull(nextDay, shift)) {
                                int si = shiftIndex(shift);
                                schedule.get(nextDay).get(si).add(emp.name);
                                employeeAssignments.computeIfAbsent(emp.name, k -> new ArrayList<>())
                                    .add(new int[] { nextDay, si });
                                daysAssigned++;
                                break;
                            }
                        }
                    }
                }
            }
        }
    }

    private void backfillShifts(List<Employee> employees) {
        List<String> names = new ArrayList<>();
        for (Employee e : employees) names.add(e.name);

        boolean added = true;
        while (added) {
            added = false;
            buildEmployeeAssignmentsFromSchedule();

            for (int day = 0; day < 7; day++) {
                for (int si = 0; si < SHIFTS.length; si++) {
                    int current = schedule.get(day).get(si).size();
                    int needed = Math.max(0, MIN_EMPLOYEES_PER_SHIFT - current);
                    if (needed == 0) continue;

                    List<String> candidates = new ArrayList<>();
                    for (String n : names) {
                        if (schedule.get(day).get(si).contains(n)) continue;
                        if (daysWorked(n) >= MAX_DAYS_PER_EMPLOYEE) continue;
                        if (isAssignedOnDay(n, day)) continue;
                        candidates.add(n);
                    }
                    Collections.shuffle(candidates, RANDOM);
                    for (int i = 0; i < Math.min(needed, candidates.size()); i++) {
                        String name = candidates.get(i);
                        schedule.get(day).get(si).add(name);
                        employeeAssignments.computeIfAbsent(name, k -> new ArrayList<>())
                            .add(new int[] { day, si });
                        added = true;
                        break;
                    }
                }
            }
        }
    }

    /** Returns sample employee data; also used by GUI to load sample data. */
    public static List<Employee> collectInput() {
        List<Employee> employees = new ArrayList<>();

        Map<Integer, String> prefsAlice = new HashMap<>();
        prefsAlice.put(0, "morning"); prefsAlice.put(1, "morning"); prefsAlice.put(2, "afternoon");
        prefsAlice.put(3, "evening"); prefsAlice.put(4, "morning"); prefsAlice.put(5, "afternoon"); prefsAlice.put(6, "evening");
        Map<Integer, List<String>> prioAlice = new HashMap<>();
        prioAlice.put(0, Arrays.asList("morning", "afternoon", "evening"));
        prioAlice.put(1, Arrays.asList("morning", "evening", "afternoon"));
        prioAlice.put(2, Arrays.asList("afternoon", "morning", "evening"));
        prioAlice.put(3, Arrays.asList("evening", "afternoon", "morning"));
        prioAlice.put(4, Arrays.asList("morning", "afternoon", "evening"));
        prioAlice.put(5, Arrays.asList("afternoon", "morning", "evening"));
        prioAlice.put(6, Arrays.asList("evening", "morning", "afternoon"));
        employees.add(new Employee("Alice", prefsAlice, prioAlice));

        Map<Integer, String> prefsBob = new HashMap<>();
        prefsBob.put(0, "afternoon"); prefsBob.put(1, "evening"); prefsBob.put(2, "morning");
        prefsBob.put(3, "morning"); prefsBob.put(4, "afternoon"); prefsBob.put(5, "evening"); prefsBob.put(6, "morning");
        Map<Integer, List<String>> prioBob = new HashMap<>();
        for (int d = 0; d < 7; d++) {
            if (d % 3 == 0) prioBob.put(d, Arrays.asList("afternoon", "evening", "morning"));
            else if (d % 3 == 1) prioBob.put(d, Arrays.asList("evening", "morning", "afternoon"));
            else prioBob.put(d, Arrays.asList("morning", "afternoon", "evening"));
        }
        employees.add(new Employee("Bob", prefsBob, prioBob));

        Map<Integer, String> prefsCarol = new HashMap<>();
        prefsCarol.put(0, "evening"); prefsCarol.put(1, "afternoon"); prefsCarol.put(2, "evening");
        prefsCarol.put(3, "afternoon"); prefsCarol.put(4, "evening"); prefsCarol.put(5, "morning"); prefsCarol.put(6, "afternoon");
        Map<Integer, List<String>> prioCarol = new HashMap<>();
        for (int d = 0; d < 7; d++) prioCarol.put(d, Arrays.asList("evening", "afternoon", "morning"));
        employees.add(new Employee("Carol", prefsCarol, prioCarol));

        Map<Integer, String> prefsDavid = new HashMap<>();
        prefsDavid.put(0, "morning"); prefsDavid.put(1, "afternoon"); prefsDavid.put(2, "morning");
        prefsDavid.put(3, "evening"); prefsDavid.put(4, "afternoon"); prefsDavid.put(5, "morning"); prefsDavid.put(6, "evening");
        employees.add(new Employee("David", prefsDavid, null));

        Map<Integer, String> prefsEve = new HashMap<>();
        prefsEve.put(0, "afternoon"); prefsEve.put(1, "morning"); prefsEve.put(2, "evening");
        prefsEve.put(3, "morning"); prefsEve.put(4, "evening"); prefsEve.put(5, "afternoon"); prefsEve.put(6, "morning");
        Map<Integer, List<String>> prioEve = new HashMap<>();
        for (int d = 0; d < 7; d++) prioEve.put(d, Arrays.asList("afternoon", "morning", "evening"));
        employees.add(new Employee("Eve", prefsEve, prioEve));

        Map<Integer, String> prefsFrank = new HashMap<>();
        prefsFrank.put(0, "morning"); prefsFrank.put(1, "evening"); prefsFrank.put(2, "afternoon");
        prefsFrank.put(3, "morning"); prefsFrank.put(4, "morning"); prefsFrank.put(5, "evening"); prefsFrank.put(6, "afternoon");
        employees.add(new Employee("Frank", prefsFrank, null));

        Map<Integer, String> prefsGrace = new HashMap<>();
        prefsGrace.put(0, "evening"); prefsGrace.put(1, "afternoon"); prefsGrace.put(2, "morning");
        prefsGrace.put(3, "evening"); prefsGrace.put(4, "afternoon"); prefsGrace.put(5, "morning"); prefsGrace.put(6, "evening");
        Map<Integer, List<String>> prioGrace = new HashMap<>();
        for (int d = 0; d < 7; d++) prioGrace.put(d, Arrays.asList("evening", "afternoon", "morning"));
        employees.add(new Employee("Grace", prefsGrace, prioGrace));

        Map<Integer, String> prefsHenry = new HashMap<>();
        prefsHenry.put(0, "afternoon"); prefsHenry.put(1, "morning"); prefsHenry.put(2, "evening");
        prefsHenry.put(3, "afternoon"); prefsHenry.put(4, "evening"); prefsHenry.put(5, "afternoon"); prefsHenry.put(6, "morning");
        employees.add(new Employee("Henry", prefsHenry, null));

        Map<Integer, String> prefsIvy = new HashMap<>();
        prefsIvy.put(0, "morning"); prefsIvy.put(1, "evening"); prefsIvy.put(2, "morning");
        prefsIvy.put(3, "afternoon"); prefsIvy.put(4, "afternoon"); prefsIvy.put(5, "morning"); prefsIvy.put(6, "evening");
        Map<Integer, List<String>> prioIvy = new HashMap<>();
        for (int d = 0; d < 7; d++) prioIvy.put(d, Arrays.asList("morning", "evening", "afternoon"));
        employees.add(new Employee("Ivy", prefsIvy, prioIvy));

        Map<Integer, String> prefsJack = new HashMap<>();
        prefsJack.put(0, "evening"); prefsJack.put(1, "morning"); prefsJack.put(2, "afternoon");
        prefsJack.put(3, "evening"); prefsJack.put(4, "morning"); prefsJack.put(5, "afternoon"); prefsJack.put(6, "morning");
        employees.add(new Employee("Jack", prefsJack, null));

        Map<Integer, String> prefsKate = new HashMap<>();
        prefsKate.put(0, "afternoon"); prefsKate.put(1, "evening"); prefsKate.put(2, "morning");
        prefsKate.put(3, "morning"); prefsKate.put(4, "evening"); prefsKate.put(5, "evening"); prefsKate.put(6, "afternoon");
        Map<Integer, List<String>> prioKate = new HashMap<>();
        for (int d = 0; d < 7; d++) prioKate.put(d, Arrays.asList("afternoon", "evening", "morning"));
        employees.add(new Employee("Kate", prefsKate, prioKate));

        return employees;
    }

    private static String repeat(char c, int n) {
        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; i++) sb.append(c);
        return sb.toString();
    }

    private void printSchedule() {
        System.out.println(formatScheduleToString());
    }

    /** Returns the schedule as a formatted string (for GUI or other use). */
    public String formatScheduleToString() {
        StringBuilder sb = new StringBuilder();
        sb.append("\n").append(repeat('=', 70)).append("\n");
        sb.append("                    FINAL EMPLOYEE SCHEDULE - WEEK\n");
        sb.append(repeat('=', 70)).append("\n");
        for (int day = 0; day < 7; day++) {
            sb.append("\n  ").append(DAYS[day].toUpperCase()).append("\n");
            sb.append("  ").append(repeat('-', 60)).append("\n");
            for (int si = 0; si < SHIFTS.length; si++) {
                List<String> names = schedule.get(day).get(si);
                String namesStr = names.isEmpty() ? "(none)" : String.join(", ", names);
                String shiftName = SHIFTS[si].substring(0, 1).toUpperCase() + SHIFTS[si].substring(1);
                sb.append(String.format("    %-10s : %s%n", shiftName, namesStr));
            }
        }
        sb.append("\n").append(repeat('=', 70)).append("\n");
        return sb.toString();
    }

    public void run() {
        List<Employee> employees = collectInput();
        boolean usePriority = false;
        for (Employee e : employees) {
            if (e.priorityRanking != null && !e.priorityRanking.isEmpty()) {
                usePriority = true;
                break;
            }
        }
        firstPassAssign(employees, usePriority);
        backfillShifts(employees);
        printSchedule();
    }

    /** Runs scheduling with sample data and returns the schedule as a string (for GUI). */
    public static String runAndGetScheduleOutput() {
        return runAndGetScheduleOutput(collectInput());
    }

    /** Runs scheduling with the given employees and returns the schedule as a string (for GUI). */
    public static String runAndGetScheduleOutput(List<Employee> employees) {
        if (employees == null || employees.isEmpty()) return "Add at least one employee.";
        ScheduleManager manager = new ScheduleManager();
        boolean usePriority = false;
        for (Employee e : employees) {
            if (e.priorityRanking != null && !e.priorityRanking.isEmpty()) {
                usePriority = true;
                break;
            }
        }
        manager.firstPassAssign(employees, usePriority);
        manager.backfillShifts(employees);
        return manager.formatScheduleToString();
    }

    public static void main(String[] args) {
        if (args.length > 0 && "gui".equalsIgnoreCase(args[0])) {
            ScheduleManagerGUI.launch();
            return;
        }
        ScheduleManager manager = new ScheduleManager();
        manager.run();
    }
}
